### PR TITLE
Fix #4724 (delegating creator, Records)

### LINKED ISSUE #4724: Deserialization behavior change with Records, `@JsonCreator` and `@JsonValue` between 2.17 and 2.18
### Search before asking

- [X] I searched in the [issues](https://github.com/FasterXML/jackson-databind/issues) and found nothing similar.

### Describe the bug

Using both JsonCreator and JsonValue annotations with Java Records makes ObjectMapper seemingly skip single-arg static JsonCreator method and try to contract instance directly. I couldn't find any hint that this was a planned change.

### Version Information

2.18.0

### Reproduction

The following code works with 2.17.2, but fails with 2.18.0

```java
    @Test
    void deserialization() throws Exception {
        new ObjectMapper().readValue("\"\"", Something.class);
    }

    public record Something(String value) {

        public Something {
            if (value == null || value.isEmpty()) {
                throw new IllegalArgumentException("Value cannot be null or empty");
            }
        }

        @JsonCreator
        public static Something of(String value) {
            if (value.isEmpty()) {
                return null;
            }
            return new Something(value);
        }

        @Override
        @JsonValue
        public String toString() {
            return value;
        }

    }

``` 

Removing JsonValue annotation makes deserialization to work again.

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by cowtowncoder**

### CODE REVIEW

**Comment by cowtowncoder** [comment]
@JooHyukKim I think this is a bit better way to fix #4724; fortunately #4731 helped a lot in figuring out what is actually going on. I am not super happy with the block of code I changed but it has to do for now.

**Comment by cowtowncoder** [comment]
Ok, did tiny bit of refactoring to -- I hope -- streamline code a bit, to make it bit easier to follow.

File: src/test-jdk17/java/com/fasterxml/jackson/databind/tofix/RecordJsonCreatorAndJsonValue4724Test.java [hunk 1]
```diff
@@ -1,24 +1,23 @@
...
+
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonValue;
 import com.fasterxml.jackson.databind.testutil.DatabindTestUtil;
-import com.fasterxml.jackson.databind.testutil.failure.JacksonTestFailureExpected;
-
-import org.junit.jupiter.api.Test;
 
 // [databind#4724] Deserialization behavior change with Java Records, JsonCreator and JsonValue between 2.17.2 => 2.18.0
 public class RecordJsonCreatorAndJsonValue4724Test
```
Comment (JooHyukKim) [inline src/test-jdk17/java/com/fasterxml/jackson/databind/tofix/RecordJsonCreatorAndJsonValue4724Test.java:10]: I was thinking we should start building habit of moving fixed `/tofix` tests out of `/tofix` package :))

Comment (cowtowncoder) [reply src/test-jdk17/java/com/fasterxml/jackson/databind/tofix/RecordJsonCreatorAndJsonValue4724Test.java:10]: We do, but I'll do that after merging PR -- there's merge conflict when moving.

**Pull request merged**